lib/worker.go

```diff
@@ -0,0 +1,434 @@
+					log.Error(err.Error())
+					continue
+				}
+				if err := w.queue.Send(tg); err != nil {
+					log.Error(err.Error())
+					continue
+				}
+			case <-w.stop:
+				w.queue.Close()
+				break
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Stop the worker
+func (w *LocalWorker) Stop() error {
+	w.stop <- true
+	return nil
+}
+
+// newMux creates a muxer (response multiplexer) for a fetchbot
+func newMux(coord WorkCoordinator, recordRedirects, recordHeaders bool) *fetchbot.Mux {
+	// Create the muxer
+	mux := fetchbot.NewMux()
+
+	// Handle all errors the same
+	mux.HandleErrors(fetchbot.HandlerFunc(func(ctx *fetchbot.Context, res *http.Response, err error) {
+		log.Infof("[ERR] %s %s - %s", ctx.Cmd.Method(), ctx.Cmd.URL(), err.Error())
+		coord.Completed(&Resource{Error: err.Error()})
+		return
+	}))
+
+	// Handle GET requests for html responses, to parse the body and enqueue all links as HEAD requests.
+	mux.Response().Method("GET").Handler(fetchbot.HandlerFunc(
+		func(ctx *fetchbot.Context, res *http.Response, err error) {
+
+			r := &Resource{URL: ctx.Cmd.URL().String()}
+
+			// TODO - huh? why this here? figure out & make a note
+			if recordRedirects {
+				r = &Resource{URL: NormalizeURLString(res.Request.URL)}
+			}
+
+			log.Infof("[%d] %s %s", res.StatusCode, ctx.Cmd.Method(), r.URL)
+
+			var st time.Time
+			if timedCmd, ok := ctx.Cmd.(*TimedCmd); ok {
+				st = timedCmd.Started
+			}
+
+			if err := r.HandleResponse(st, res, recordHeaders); err != nil {
+				log.Debugf("error handling get response: %s - %s", ctx.Cmd.URL().String(), err.Error())
+				return
+			}
+
+			if err := coord.Completed(r); err != nil {
+				log.Errorf("[ERR] coordinator: %s", err.Error())
+			}
+		}))
+
+	return mux
+}
+
+// stopHandler stops the fetcher if the stopurl is reached. Otherwise it dispatches
+// the call to the wrapped Handler.
+func stopHandler(stopurl string, stop chan bool, wrapped fetchbot.Handler) fetchbot.Handler {
+	return fetchbot.HandlerFunc(func(ctx *fetchbot.Context, res *http.Response, err error) {
+		if ctx.Cmd.URL().String() == stopurl {
+			log.Infof(">>>>> STOP URL %s\n", ctx.Cmd.URL())
+			// generally not a good idea to stop/block from a handler goroutine
+			// so do it in a separate goroutine
+			go func() {
+				stop <- true
+			}()
+			return
+		}
+		wrapped.Handle(ctx, res, err)
+	})
+}
+
+// NewRecordRedirectClient creates a http client with a custom checkRedirect function that
+// creates records of Redirects & sends them to the coordinator
+func NewRecordRedirectClient(wc WorkCoordinator) *http.Client {
+	return &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+
+			prev := via[len(via)-1]
+			prevurl := NormalizeURLString(prev.URL)
+
+			r, _ := url.Parse(req.URL.String())
+			canurlstr := NormalizeURLString(r)
+
+			if prevurl != canurlstr {
+				log.Infof("[%d] %s %s -> %s", req.Response.StatusCode, prev.Method, prevurl, canurlstr)
+
+				wc.Completed(&Resource{
```

Comment:
Is this call going to potentially cause the coordinator to stop early? (If there are no more items in the queue, this marks the URL the coordinator queued as done, even though there are more requests in the redirect chain here to be made.)

Reply:
Yes. This is bad. Currently I haven't run any tests with redirects, and intended to get a PR going after this that refactors redirects to `[]string`, and adds tests.